### ISSUE TITLE
SQL Server: When retrieving generated columns in insert, concurrency may cause us not to get the original generated value

### ISSUE THREAD

**Issue by roji** [dotnet/efcore#27446]
When inserting a new entity instance which has both identity and other generated values, we current use the INSERT+SELECT pattern:

```sql
INSERT INTO [Blogs] ([Name])
VALUES (@p0);
SELECT [Id], [Foo]
FROM [Blogs]
WHERE @@ROWCOUNT = 1 AND [Id] = scope_identity();
```

Although the two statements are wrapped in a transaction, we use the default isolation level, which is read committed; this means that  if an update happens between the INSERT and the SELECT, the SELECT returns the updated value for Foo, and not the originally-generated value.

I'm not sure this is an actual problem, but it does seem like we should ideally propagate back the original generated values, and not be sensitive to a race condition such as this. Switching to a single-command method (`SELECT ... OUTPUT`) would remove this problem, but would degrade performance when triggers are present (`SELECT ... OUTPUT INTO`) - see #27372 for more info.

**Comment by roji** [comment]
Design decision: we will move away from doing INSERT+SELECT. See #27372 for further decisions on insertion behavior.

**Comment by GSPP** [comment]
On SQL Server, this should not be a problem because the `INSERT` X-locks the row.

The switch to another strategy might still be wise, just mentioning it.

**Comment by roji** [comment]
@GSPP you're right...

In addition, any other transaction wouldn't even see the newly-inserted row before its SELECT completes, since the INSERT+SELECT hasn't been committed yet (unless that other transaction is using READ UNCOMMITTED). In fact

So I agree that there isn't a correctness issue here, and if we want to keep perf optimized for all scenarios (including trigger-enabled ones), we may want to keep this technique. @GSPP any other reason you see to switch away from this?

Note: in the context of avoiding transactions around single updates (#27439), this is a good example of a single modification command which is executed via 2 SQL commands, which do need to be wrapped in a transaction.

/cc @lauxjpn

**Comment by GSPP** [comment]
> deadlock by doing INSERT+SELECT in one transaction and UPDATE in another

How do you see that deadlock happening? I can't see it.

<hr>

Using a single DML statement that outputs the data that the client needs *feels* right to me. It can't get better than this. The only downside is that `MERGE` might not collapse to the exact same efficiency as an `INSERT`. It should be pretty close, though. I see no reason why there would be any difference at all, but you never know.

To me, these idiosyncrasies in SQL Server with the various strategies seem a bit unnecessary. They seem like unforced design errors in SQL Server. We have to make do with what we've got.

**Comment by roji** [comment]
>> deadlock by doing INSERT+SELECT in one transaction and UPDATE in another
>
> How do you see that deadlock happening? I can't see it.

I didn't give enough details above; I could reproduce a deadlock when the INSERT+SELECT transaction waits between the two commands on an unrelated UPDATE. So INSERT acquires the lock as you wrote above, then a separate UPDATE transaction blocks on that lock, as expected. The only weirdness IMHO is that this also happens in RSCI mode, where the separate UPDATE transaction isn't even supposed to see the inserted row, and so not block on it (this is the PG behavior).

But I'm not SQL Server expert, so any insights would be appreciated. Here's a code sample:

<details>
<summary>Deadlock INSERT+UPDATE</summary>

```c#
await using var conn1 = new SqlConnection("Server=localhost;Database=test;User=SA;Password=Abcd5678;Connect Timeout=60;ConnectRetryCount=0;Encrypt=false");
await conn1.OpenAsync();

await using var conn2 = new SqlConnection("Server=localhost;Database=test;User=SA;Password=Abcd5678;Connect Timeout=60;ConnectRetryCount=0;Encrypt=false");
await conn2.OpenAsync();

await using var cmd1 = new SqlCommand();
cmd1.Connection = conn1;
cmd1.CommandText = @"
ALTER DATABASE test SET READ_COMMITTED_SNAPSHOT ON;
DROP TABLE IF EXISTS Foo;
CREATE TABLE Foo (Id INT IDENTITY PRIMARY KEY, Bar INT)";
await cmd1.ExecuteNonQueryAsync();

await using var tx = (SqlTransaction)await conn1.BeginTransactionAsync();
cmd1.Transaction = tx;
cmd1.CommandText = @"INSERT INTO Foo (Bar) VALUES (8)";
await cmd1.ExecuteNonQueryAsync();

await using var cmd2 = new SqlCommand("UPDATE Foo SET Bar = 1000", conn2);
await cmd2.ExecuteNonQueryAsync();

cmd1.CommandText = @"SELECT [Bar] FROM [Foo] WHERE @@ROWCOUNT = 1 AND [Id] = scope_identity()";
Console.WriteLine(await cmd1.ExecuteScalarAsync());
```

</details>

> Using a single DML statement that outputs the data that the client needs feels right to me. It can't get better than this. The only downside is that MERGE might not collapse to the exact same efficiency as an INSERT. It should be pretty close, though. I see no reason why there would be any difference at all, but you never know.

I certainly agree, and what you write is true - except if there happen to be triggers on the table. This is because you can't use `INSERT ... OUTPUT` (or `MERGE ... OUTPUT`) without the INTO clause (using a TVP), and that is the source of considerable slowdown. So for the trigger case, doing INSERT+SELECT does make sense from a perf perspective, for a small number of rows (before `MERGE ... OUTPUT INTO` becomes faster).

**Comment by lauxjpn** [comment]
> But I'm not SQL Server expert, so any insights would be appreciated. Here's a code sample:

I think the code sample might not be accurate in regards to investigating a potential deadlock issue. I think it should look something like the following instead (check me on that):

<details>
<summary>Program.cs</summary>

```c#
using Microsoft.Data.SqlClient;

namespace MssqlSnapshotIsolation01;

internal static class Program
{
    private static async Task Main()
    {
        await using (var conn0 = new SqlConnection(@"Server=.\MSSQL14;Integrated Security=True;Encrypt=False"))
        {
            await conn0.OpenAsync();
                
            await using var cmd0 = new SqlCommand();
            cmd0.Connection = conn0;
            cmd0.CommandText = @"
    DROP DATABASE IF EXISTS [EFCoreIssue27446_01];
    CREATE DATABASE [EFCoreIssue27446_01];
    ALTER DATABASE [EFCoreIssue27446_01] SET ALLOW_SNAPSHOT_ISOLATION ON;
    ALTER DATABASE [EFCoreIssue27446_01] SET READ_COMMITTED_SNAPSHOT ON;
    USE [EFCoreIssue27446_01];
    CREATE TABLE Foo (Id INT IDENTITY PRIMARY KEY, Bar INT)";
                
            await cmd0.ExecuteNonQueryAsync();
        }

        await using var conn1 = new SqlConnection(@"Server=.\MSSQL14;Integrated Security=True;Encrypt=False;Database=EFCoreIssue27446_01");
        await conn1.OpenAsync();

        await using var conn2 = new SqlConnection(@"Server=.\MSSQL14;Integrated Security=True;Encrypt=False;Database=EFCoreIssue27446_01");
        await conn2.OpenAsync();

        await using var tx1 = (SqlTransaction)await conn1.BeginTransactionAsync();
        await using var cmd1 = new SqlCommand();
        cmd1.Connection = conn1;
        cmd1.Transaction = tx1;
        cmd1.CommandText = @"INSERT INTO Foo (Bar) VALUES (8)";
        var task1 = cmd1.ExecuteNonQueryAsync();

        await using var cmd2 = new SqlCommand("UPDATE Foo SET Bar = 1000", conn2);
        var task2 = cmd2.ExecuteNonQueryAsync();

        await task1;
            
        cmd1.CommandText = @"SELECT [Bar] FROM [Foo] WHERE @@ROWCOUNT = 1 AND [Id] = scope_identity()";
        Console.WriteLine(await cmd1.ExecuteScalarAsync());

        await tx1.CommitAsync();
        
        await task2;
    }
}
```

</details>

This works as expected (without deadlock).

**Comment by roji** [comment]
@lauxjpn thanks for looking into this!

Regarding locking and RCSI:

> With the write lock in mind for RCSI, the code sample would need to be change to something like the following to work as expected [...]

I think I'm probably missing something... Compared to my code sample above, your code gives UPDATE 8 seconds of waiting before continuing to do the SELECT and commit the transaction. Unless I'm mistaken, that means that the UPDATE is still blocking, waiting for the X-Lock from the INSERT transaction to be released. After the 8 seconds wait, your code commits the INSERT transaction and so the UPDATE is released.

In other words, my point above was only to point out that UPDATE does see the row and lock it, although the transaction creating it hasn't committed yet - and all that with RCSI turned on, and with READ COMMITTED (not READ UNCOMMITTED). That's the weird part for me, and on PG it does not occur.

But this has no actual bearing on this discussion on whether INSERT+SELECT is safe: the only important point is that as long as INSERT+SELECT are wrapped in a transaction, an external update cannot interfere with the values the SELECT sees (on SQL Server).

**Comment by roji** [comment]
Closing; as discussed above, there's no safety issue with INSERT+SELECT as long as they're wrapped in a transaction (thanks @GSPP). We'll keep this behavior for trigger mode.
